staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/dynamic_cafile_content.go

```diff
@@ -158,17 +159,51 @@ func (c *DynamicFileCAContent) Run(workers int, stopCh <-chan struct{}) {
 	// doesn't matter what workers say, only start one.
 	go wait.Until(c.runWorker, time.Second, stopCh)
 
-	// start timer that rechecks every minute, just in case.  this also serves to prime the controller quickly.
-	go wait.PollImmediateUntil(FileRefreshDuration, func() (bool, error) {
-		c.queue.Add(workItemKey)
-		return false, nil
-	}, stopCh)
-
-	// TODO this can be wired to an fsnotifier as well.
+	// start the loop that watches the CA file until stopCh is closed.
+	go wait.Until(func() {
+		if err := c.watchCAFile(stopCh); err != nil {
+			klog.ErrorS(err, "Failed to watch CA file, will retry later")
+		}
+	}, time.Minute, stopCh)
 
 	<-stopCh
 }
 
+func (c *DynamicFileCAContent) watchCAFile(stopCh <-chan struct{}) error {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error creating fsnotify watcher: %v", err)
+	}
+	defer w.Close()
+
+	if err = w.Add(c.filename); err != nil {
+		return fmt.Errorf("error adding watch for file %s: %v", c.filename, err)
+	}
+	// Reload the content when it starts a new watch as the file may have changed.
+	c.queue.Add(workItemKey)
```

Comment:
Please do this first in your method, so you will always trigger a check of the content when called.  Otherwise an error above could prevent the call.

Reply:
> And if the above returns an error, it should be because the file is missing so checking the content should fail too, do you think it's necessary to trigger it in the begining and after starting watch?

Yes, I think that's better.  I want to be sure we always trigger.  It effectively makes this method a "fire every minute" in the worst case and "fire exactly when needed" in the good cases.

Reply:
done